### PR TITLE
feat: add AssetManager

### LINKED ISSUE #602: Migrate AssetManager to @dfinity/assets
Currently there's the `AssetManager` implementation available [here](https://www.npmjs.com/package/@slide-computer/assets).

This issue is regarding improving on this implementation and moving it to `@dfinity/assets`.

The `AssetManager` has `list`, `store`, `delete`, `get`, `batch` and `clear` methods to manage assets in an asset storage canister.

### PR DESCRIPTION
**Pull request by sea-snake**
# Description

Currently there's the AssetManager implementation available [here](https://www.npmjs.com/package/@slide-computer/assets).

This PR improves on this implementation and moves it to @dfinity/assets.

The AssetManager has list, store, delete, get, batch and clear methods to manage assets in an asset storage canister.

Fixes #602

# How Has This Been Tested?

- `AssetManager` has e2e tests (Node.js)
- `Readable` has unit tests

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by sea-snake** [comment]
PR is work in progress.

Main points that are still TODO:
- Implement a way to track events and/or progress
- Additional unit and e2e tests
- Documentation
- Example application

**Comment by sea-snake** [comment]
Example application PR: https://github.com/dfinity/examples/pull/355

File: packages/assets/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+    const chunks = Array.from<number[]>({ length: Math.ceil(this.length / this.chunkSize) });
+    await this.getChunks((index, chunk) => (chunks[index] = Array.from(chunk)));
+    return chunks.flat();
+  }
+
+  /**
+   * Write asset content to file (Node.js)
+   */
+  public async write(path: string): Promise<void> {
+    const fd = await new Promise<number>(async (resolve, reject) =>
```
Comment (krpeacock) [inline packages/assets/src/index.ts:431]: This will need to be refactored for the lint check

File: packages/assets/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+    let certificate: ArrayBuffer | undefined;
+    let tree: ArrayBuffer | undefined;
+    let encoding = '';
+    for (const [key, value] of response.headers) {
+      switch (key.trim().toLowerCase()) {
+        case 'ic-certificate':
+          {
+            const fields = value.split(/,/);
+            for (const f of fields) {
+              // @ts-ignore
```
Comment (krpeacock) [inline packages/assets/src/index.ts:510]: we prefer `as unknown as <expected-type>` when the TS engine can't pick up on an expected type

File: packages/assets/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "allowJs": false,
```
Comment (krpeacock) [inline packages/assets/tsconfig.json:4]: if you do choose to make the IDL a JS file, you'll need to set this to true

**Review by krpeacock — CHANGES_REQUESTED** [review]

**Comment by krpeacock** [comment]
This is close to ready, and the main logic looks solid! To manually run the lint checks, you can run `npm run lint --workspace @dfinity/assets`

**Comment by sea-snake** [comment]
> This is close to ready, and the main logic looks solid! To manually run the lint checks, you can run `npm run lint --workspace @dfinity/assets`

Thanks for the helpful feedback! I've resolved these points in the latest commit.

**Review by krpeacock — APPROVED** [review]

**Comment by domwoe** [comment]
@sea-snake Seems there is still an issue with the timeout for the E2E tests.

**Comment by sea-snake** [comment]
> @sea-snake Seems there is still an issue with the timeout for the E2E tests.

Ended up reducing the asset size in the tests (e.g. from 1MB to 1KB) to decrease the test duration. On a local environment, the speed in more in line with expectation (2s) for each chunk but it seems things take a bit longer on the automated test env.

**Comment by domwoe** [comment]
Thanks @sea-snake!
@krpeacock Seems to be ready to be merged now.

**Comment by krpeacock** [comment]
@sea-snake is it concerning that the tests were timing out with a minute to run for a 3mb upload?

**Comment by sea-snake** [comment]
@krpeacock 
I've benchmarked all steps involved. Some observations I've made:
- Sha256 in JS is slow, sadly the current native web implementation doesn't allow incremental hashing so it's not an alternative (You don't want to load a big file in memory if avoidable).
- Every update call takes approx. 2s. For a 3MB asset test there are 5 calls (create_batch, create_chunk, create_chunk, commit_batch and delete_asset). If all these calls were made sequentially, this would be 10s. The create_chunk calls are made in parallel in theory but the dfx environment (local, production, system it's running on) might make a difference there.
- The psuedo random bytes generator in the test isn't the fastest code out there but didn't seem to make a big influence on the overall test duration.
- Maybe the cbor encoding of the update call messages might also affect the speed depending on the system it's running on.

**Comment by sea-snake** [comment]
@krpeacock 
One difference between node and web is that fetch in node is still using http1.1 instead of http2 if I remember correctly. Might be interesting to see if there's a difference in performance when tests are run in cypress.

**Comment by krpeacock** [comment]
I'll go ahead and merge so we can get this work in front of people, and I'll create an issue to follow up on automating some benchmarks

**Comment by sea-snake** [comment]
Doing some more debugging now to see if there are some other issues slowing things down.

**Pull request merged**